corehq/apps/userreports/adapter.py

```diff
@@ -101,6 +103,44 @@ def run_asynchronous(self):
     def get_distinct_values(self, column, limit):
         raise NotImplementedError
 
+    def log_table_build(self, initiated_by, source):
+        from corehq.apps.userreports.models import DataSourceActionLog
+        self._log_action(initiated_by, source, DataSourceActionLog.BUILD)
+
+    def log_table_rebuild(self, initiated_by, source, skip=False):
+        from corehq.apps.userreports.models import DataSourceActionLog
+        self._log_action(initiated_by, source, DataSourceActionLog.REBUILD, skip=skip)
+
+    def log_table_drop(self, initiated_by, source, skip=False):
+        from corehq.apps.userreports.models import DataSourceActionLog
+        self._log_action(initiated_by, source, DataSourceActionLog.DROP, skip=skip)
+
+    def log_table_migrate(self, source, diffs):
+        from corehq.apps.userreports.models import DataSourceActionLog
+        self._log_action(None, source, DataSourceActionLog.MIGRATE, diffs=diffs)
+
+    def _log_action(self, initiated_by, source, action, diffs=None, skip=False):
+        from corehq.apps.userreports.models import DataSourceActionLog
+        if skip or not self.config.data_source_id:
+            return
+
+        kwargs = {
+            'domain': self.config.domain,
+            'indicator_config_id': self.config.data_source_id,
+            'action': action,
+            'initiated_by': initiated_by,
+            'action_source': source,
+            'migration_diffs': diffs
+        }
+
+        try:
+            # make this atomic so that errors don't affect outer transactions
+            with transaction.atomic():
+                DataSourceActionLog.objects.create(**kwargs)
+        except:  # noqa
```

Comment:
What do you think about adding an exception handler above this?
```py
except (KeyboardInterrupt, SystemExit):
    raise
except:  # noqa
    ...
```

Silencing exceptions like that causes usability issues when running code from the command line (and then subsequently trying to kill it with Ctrl+C) or when debugging with (temporary) code like `sys.exit()` to force the program to stop immediately.

Reply:
https://github.com/dimagi/commcare-hq/pull/24196